### ISSUE TITLE
Warn about non-printable characters in string literals

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust#63682]
Currently a string literal with control characters like `\0` or `\v` is accepted without any warnings. The only exception is `\r`, which gives a hard error.

It makes more sense to treat all non-[printable](https://en.wikipedia.org/wiki/ASCII#Printable_characters), non \t, \n ASCII characters as a warning. 

Steps to fix:

1. Add  `NonPrintableAsii` to [EscapeError](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/librustc_lexer/src/unescape.rs#L11)
2. Produce this error somewhere around [here](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/librustc_lexer/src/unescape.rs#L135-L137)
3. Add lexer-level [tests](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/librustc_lexer/src/unescape/tests.rs)
4. Handle this "error" in [`unespcape_error_reporting`](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/libsyntax/parse/unescape_error_reporting.rs#L38). Note that, unlike other real errors, this one should be just a warning.
5. Adjust the affected ui tests

I am not sure how to make this warning work with `#[allow]` lint infrastructure: we definitely can't do this in the lexer.

**Comment by matklad** [comment]
cc @petrochenkov 

@rustbot modify labels: +A-parser, +A-diagnostics, +E-medium

**Comment by petrochenkov** [comment]
>I am not sure how to make this warning work with #[allow] lint infrastructure: we definitely can't do this in the lexer.

Attach the lint to `CRATE_NODE_ID`, then it'll be able to be enabled/disabled at crate level, but not with higher granularity.

**Comment by Areredify** [comment]
I'm not sure how to handle this: callers assume that errors `unescape_str` produces are hard errors, so if we don't `span_err` on it, that inconsistency causes an ICE later in type deduction (maybe inconsistent array length?). At least that what I think is happening, not entirely sure. Should I change output type to something like Ok(c) | Warn(c) | Err?

**Comment by matklad** [comment]
Hm, indeed, `unescape_str` and friends interface is not fully-prepared for emitting warnings.

One approach here is to extend the interface. I think something like this will be cleanest:

```rust
struct EscapeError {
    fallback: char,
    kind: EscapeErrorKind,
}

impl EscapeErrorKind {
    fn is_warning(&self) -> bool;
}
```

That is, even if we emit error, we still produce a fallback char that should be inserted into the string.

However, this particular warning can be handled with less machinery. At the call-site, we check if `char` is an unprintable character, and range is of length one. Range length seems seems like a hacky, but correct way to figure out if the char was escaped. 

The benefit of the first approach is that it centralizes the knowledge about this warning to the `rustc_lexer` crate. The drawback is that it makes an already highly awkward interface even more cumbersome. 

The benefit of the second approach is simplicity. Moreover, it can be argued that warnings not necessary belong to rustc_lexer, as they don't affect the language definition. 

I personally would prefer to start with the second approach: it seems significantly easier to implement, and is very isolated, so changing it to the second one in the future would be easy. OTOH, starting with 2 and simplifying to 1 would be much harder to do.

**Comment by Alexendoo** [comment]
Triage: Hi, are you still working on this issue @Areredify?

**Comment by Areredify** [comment]
@Alexendoo nope, you can grab it, check out my pr earlier for reference if you want

**Comment by Areredify** [comment]
oh, sorry, it's a triage, I thought you wanted to grab it 😅

**Comment by RodBurman** [comment]
Do we actually want to do this? Make code such as  ```const RED: &str = "\x1b[38;5;3m";``` or ```const WINDOWS_EOL: &str = "\r\n"``` cause a warning? These are currently perfectly valid snippets. If there is evidence that adding \0, \r or the like without meaning to causes a common set of problems may be something like ```const NO_CTL: & str = p"printable characters only"``` which made such a promise would be useful? But I think simply adding such a non-printable warning would "break" a lot of current code without highlighting many problems.
